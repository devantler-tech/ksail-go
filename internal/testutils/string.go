package testutils

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

// AssertStringContains asserts that the haystack contains all provided substrings.
func AssertStringContains(t *testing.T, haystack string, subs ...string) {
	t.Helper()

	for _, s := range subs {
		assert.Contains(t, haystack, s)
	}
}

// AssertStringContainsOneOf asserts that the haystack contains at least one of the provided options.
func AssertStringContainsOneOf(t *testing.T, haystack string, options ...string) {
	t.Helper()

	for _, opt := range options {
		if assert.Contains(t, haystack, opt) {
			return
		}
	}

	assert.Failf(
		t,
		"string did not contain any expected option",
		"wanted one of %v in: %s",
		options,
		haystack,
	)
}
