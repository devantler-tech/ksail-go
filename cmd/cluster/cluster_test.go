package cluster //nolint:testpackage // Needs access to unexported helpers for coverage instrumentation.

import (
	"bytes"
	"errors"
	"strings"
	"testing"

	"github.com/devantler-tech/ksail-go/cmd/cluster/testutils"
	configmanager "github.com/devantler-tech/ksail-go/pkg/config-manager/ksail"
	"github.com/spf13/cobra"
)

func TestNewClusterCmdRegistersLifecycleCommands(t *testing.T) {
	t.Parallel()

	metadata := expectedLifecycleMetadata(t)
	requireParentMetadata(t, NewClusterCmd())

	for name, details := range metadata {
		t.Run(name, func(t *testing.T) {
			t.Parallel()

			subcommand := findLifecycleSubcommand(t, name)
			assertSubcommandMetadata(t, subcommand, details)
		})
	}
}

func TestClusterCommandRunEDisplaysHelp(t *testing.T) {
	t.Parallel()

	cmd := NewClusterCmd()
	buffer := &bytes.Buffer{}
	cmd.SetOut(buffer)
	cmd.SetErr(buffer)
	cmd.SetArgs(nil)

	err := cmd.Execute()
	if err != nil {
		t.Fatalf("expected executing cluster command without subcommand to succeed, got %v", err)
	}

	assertOutputContains(t, buffer.String(), "Usage:")
	assertOutputContains(t, buffer.String(), "Available Commands:")
}

//nolint:paralleltest // Alters package-level helper.
func TestHandleClusterRunEWrapsHelpError(t *testing.T) {
	originalRunner := helpRunner
	helpRunner = func(*cobra.Command) error {
		return errHelpFailure
	}

	defer func() {
		helpRunner = originalRunner
	}()

	cmd := &cobra.Command{Use: "cluster"}

	err := handleClusterRunE(cmd, nil, nil)
	if !errors.Is(err, errHelpFailure) {
		t.Fatalf("expected wrapped help failure error, got %v", err)
	}
}

var errHelpFailure = errors.New("help failure")

type lifecycleMetadata struct {
	short string
	long  string
}

func expectedLifecycleMetadata(t *testing.T) map[string]lifecycleMetadata {
	t.Helper()

	constructors := []func() *cobra.Command{
		NewUpCmd,
		NewDownCmd,
		NewStartCmd,
		NewStopCmd,
		NewStatusCmd,
		NewListCmd,
	}

	metadata := make(map[string]lifecycleMetadata, len(constructors))

	for _, constructor := range constructors {
		cmd := constructor()
		metadata[cmd.Use] = lifecycleMetadata{
			short: cmd.Short,
			long:  cmd.Long,
		}
	}

	return metadata
}

func requireParentMetadata(t *testing.T, cmd *cobra.Command) {
	t.Helper()

	const expectedDescription = "Manage cluster lifecycle"

	if cmd.Short != expectedDescription {
		t.Fatalf(
			"short description mismatch for parent command. want %q, got %q",
			expectedDescription,
			cmd.Short,
		)
	}
}

func findLifecycleSubcommand(t *testing.T, name string) *cobra.Command {
	t.Helper()

	for _, subcommand := range NewClusterCmd().Commands() {
		if subcommand.Use == name {
			return subcommand
		}
	}

	t.Fatalf("expected cluster command to include %q subcommand", name)

	return nil
}

func assertSubcommandMetadata(t *testing.T, cmd *cobra.Command, metadata lifecycleMetadata) {
	t.Helper()

	if cmd.Short != metadata.short {
		t.Fatalf(
			"short description mismatch for %q. want %q, got %q",
			cmd.Use,
			metadata.short,
			cmd.Short,
		)
	}

	if cmd.Long != metadata.long {
		t.Fatalf(
			"long description mismatch for %q. want %q, got %q",
			cmd.Use,
			metadata.long,
			cmd.Long,
		)
	}
}

func assertOutputContains(t *testing.T, output, expected string) {
	t.Helper()

	if !strings.Contains(output, expected) {
		t.Fatalf("expected output to contain %q, got %q", expected, output)
	}
}

type lifecycleHandlerFunc func(*cobra.Command, *configmanager.ConfigManager, []string) error

func runLifecycleSuccessCase(
	t *testing.T,
	use string,
	handler lifecycleHandlerFunc,
	expectedMessage string,
) {
	t.Helper()

	cmd, manager, output := testutils.NewCommandAndManager(t, use)
	testutils.SeedValidClusterConfig(manager)

	err := handler(cmd, manager, nil)
	if err != nil {
		t.Fatalf("expected no error, got %v", err)
	}

	assertOutputContains(t, output.String(), expectedMessage)
}

func runLifecycleValidationErrorCase(
	t *testing.T,
	use string,
	handler lifecycleHandlerFunc,
	expectedSubstrings ...string,
) {
	t.Helper()

	testutils.RunValidationErrorTest(t, use, func(
		cmd *cobra.Command,
		manager *configmanager.ConfigManager,
		args []string,
	) error {
		err := handler(cmd, manager, args)
		if err == nil {
			t.Fatal("expected error but got nil")
		}

		message := err.Error()
		for _, substring := range expectedSubstrings {
			if !strings.Contains(message, substring) {
				t.Fatalf("expected error message to contain %q, got %q", substring, message)
			}
		}

		return err
	})
}
