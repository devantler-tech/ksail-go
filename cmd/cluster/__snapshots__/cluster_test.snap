
[TestClusterCommandRunEDisplaysHelp - 1]
Manage lifecycle operations for local Kubernetes clusters, including provisioning, teardown, and status.

Usage:
  cluster [flags]
  cluster [command]

Available Commands:
  completion  Generate the autocompletion script for the specified shell
  connect     Connect to cluster with k9s
  create      Create a cluster
  delete      Destroy a cluster
  help        Help about any command
  info        Display cluster information
  list        List clusters
  start       Start a stopped cluster
  stop        Stop a running cluster

Flags:
  -h, --help   help for cluster

Use "cluster [command] --help" for more information about a command.

---
